### DISCUSSION TITLE
Custom ausgab objects

### DISCUSSION THREAD

**Discussion by jstask0** [nrc-cnrc/EGSnrc#1040] [Questions] [answered]
I want to create a custom ausgab object for part of a project I'm working on. I am planning to copy the `egs_radiative_splitting` ausgab object and modify it to suit my needs. I know I will have to modify the `ProcessEvent` function in the header and the "process input" portion of the source code, as well as renaming certain files and variables. Are there any other things that need to be modified (like the `Makefile`)?

I think it would be great if there were a resource that walked through the basics of how to make a custom ausgab object.

**Comment by rtownson** [comment, marked as answer]
In the `Makefile` in the ausgab object directory, you'll have to rename some things. Also, in the `HEN_HOUSE/egs++/Makefile` you'll want to add yours to the `aobject_libs` list. I would start with just doing all this renaming and then see if it compiles, before you start making changes.

**Reply by jstask0** [reply]
Thank you! I've gotten it to successfully compile. I'm sure I will have more questions as I work on editing the content.

**Reply by ftessier** [reply]
@jstask0 feel free to record the steps you followed to do this: we could publish that in the EGSnrc wiki or in the manual (with due credit, of course).

**Reply by jstask0** [reply]
Sure, I can write up a summary and post it here later this week

**Reply by jstask0** [reply]
Here's the basic process I followed to create a new custom ausgab object. I'm currently working on editing the object to perform the job I need it to.
1.	In the `HEN_HOUSE/egs++/Makefile`, there is a list called `aobject_libs`. Add the name of your custom ausgab object to the end of the list. The common naming convention is `egs_` followed by a description of what the ausgab object does. For the rest of this tutorial, the custom ausgab object will be called `egs_name`.
2.	Navigate to the `HEN_HOUSE/egs++/ausgab_objects` directory, and duplicate one of the ausgab object directories found there. The simplest one is the `egs_radiative_splitting` ausgab object.
3.	Rename the duplicated directory to `egs_name`.
4.	Open the directory and rename the `.cpp` and `.h` files to `egs_name.cpp` and `egs_name.h`.
5.	Open the `Makefile`. Replace any entries that refer to the old ausgab object with your new ausgab object name. For example, we would replace `RADIATIVE_SPLITTING` and `radiative_splitting` with `NAME` or `name` where `DEFS`, `library`, and `lib_files` are declared in the `egs_radiative_splitting Makefile`.
6.	Follow similar steps to modify the `.cpp` and `.h `files in your `egs_name` directory. Be careful to rename all instances of the old ausgab object. It might be easiest to use the find function of your coding IDE to find all instances of the old ausgab object. The format for each varies. For example, there are lines that reference `radiative_splitting`, `RADIATIVE_SPLITTING`, and `RadiativeSplitting` in the `egs_radiative_splitting.cpp` and `.h` files.
7.	Once you’ve made all edits, save the files and close them.
8.	Open a command prompt window, and navigate to the `HEN_HOUSE/egs++` directory.
9.	Type `make` into the command prompt window and hit enter. A successful compilation indicates that you have made the necessary edits to the custom ausgab object files. Now you are ready to edit the content of the ausgab object to program it to complete your objectives.

**Reply by ftessier** [reply]
Awesome, thanks!
